SimFastTiming/FastTimingCommon/src/BTLBarDeviceSim.cc

```diff
@@ -0,0 +1,120 @@
+  LightCollEff_(pset.getParameter<double>("LightCollectionEff")),
+  LightCollSlopeR_(pset.getParameter<double>("LightCollectionSlopeR")),
+  LightCollSlopeL_(pset.getParameter<double>("LightCollectionSlopeL")),
+  PDE_(pset.getParameter<double>("PhotonDetectionEff")) { }
+
+void BTLBarDeviceSim::getEventSetup(const edm::EventSetup& evs) {
+  edm::ESHandle<MTDGeometry> geom;
+  if( geomwatcher_.check(evs) || geom_ == nullptr ) {
+    evs.get<MTDDigiGeometryRecord>().get(geom);
+    geom_ = geom.product();
+  }
+  edm::ESHandle<MTDTopology> mtdTopo;
+  if ( topowatcher_.check(evs) || topo_ == nullptr ) {
+    evs.get<MTDTopologyRcd>().get(mtdTopo);
+    topo_ = mtdTopo.product();
+  }
+}
+
+void BTLBarDeviceSim::getHitsResponse(const std::vector<std::tuple<int,uint32_t,float> > &hitRefs, 
+				      const edm::Handle<edm::PSimHitContainer> &hits,
+				      mtd_digitizer::MTDSimHitDataAccumulator *simHitAccumulator,
+				      CLHEP::HepRandomEngine *hre){
+
+  //loop over sorted simHits
+  for (auto const& hitRef: hitRefs) {
+
+    const int hitidx   = std::get<0>(hitRef);
+    const uint32_t id  = std::get<1>(hitRef);
+    const MTDDetId detId(id);
+    const PSimHit &hit = hits->at( hitidx );     
+    
+    // --- Safety check on the detector ID
+    if ( detId.det()!=DetId::Forward || detId.mtdSubDetector()!=1 ) continue;
+
+    if(id==0) continue; // to be ignored at RECO level                                                              
+
+    BTLDetId btlid(detId);
+    const int boundRef = ( topo_->getMTDTopologyMode() == (int ) BTLDetId::CrysLayout::barzflat ?
+			   BTLDetId::kTypeBoundariesBarZflat[1]   :
+			   BTLDetId::kTypeBoundariesReference[1] );
+    DetId geoId = BTLDetId(btlid.mtdSide(),btlid.mtdRR(),btlid.module()+boundRef*(btlid.modType()-1),0,1);
+    const MTDGeomDet* thedet = geom_->idToDet(geoId);
+
+    if( thedet == nullptr ) {
+      throw cms::Exception("BTLBarDeviceSim") << "GeographicalID: " << std::hex
+					      << geoId.rawId()
+					      << " (" << detId.rawId()<< ") is invalid!" << std::dec
+					      << std::endl;
+    }
+    const ProxyMTDTopology& topoproxy = static_cast<const ProxyMTDTopology&>(thedet->topology());
+    const RectangularMTDTopology& topo = static_cast<const RectangularMTDTopology&>(topoproxy.specificTopology());    
+    // calculate the simhit row and column                                  
+    const auto& pentry = hit.entryPoint();
+    Local3DPoint simscaled(0.1*pentry.x(),0.1*pentry.y(),0.1*pentry.z()); // mm -> cm here is the switch
+    // translate from crystal-local coordinates to module-local coordinates to get the row and column
+    simscaled = topo.pixelToModuleLocalPoint(simscaled,btlid.row(topo.nrows()),btlid.column(topo.nrows()));
+    const auto& thepixel = topo.pixel(simscaled);
+    uint8_t row(thepixel.first), col(thepixel.second);
+
+    if( btlid.row(topo.nrows()) != row || btlid.column(topo.nrows()) != col ) {
+      edm::LogWarning("BTLBarDeviceSim")
```

Comment:
> You potentially added 57447 lines to the logs

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25384/32077/runTheMatrix-results/27434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D35+RecoFullGlobal_2023D35+HARVESTFullGlobal_2023D35/step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D35+RecoFullGlobal_2023D35+HARVESTFullGlobal_2023D35.log

This message needs to be reduced in severity (or a problem needs to be solved somewhere)

Reply:
This is a consequence of a mismatch with the BTL geometry. As soon as we fix that, the warning will not be triggered so often.

Reply:
Will the mismatch be fixed in this PR? It cannot be accepted as is.

Reply:
Yes, of course.